README.md
# First Vision

This repo contains the content for tenants of First Vision.


## Directory structure

### /docs - all markdown files are to be placed in this directory
### /assets -  static assets like image etc here
### /config/document-explorer-definition.yaml: Provide explorer structure for dev studio    
### /config/tenant_api.json: Tenant Provider API 
### /config/product_layout.yaml: Yaml spec for product layout page
### /reference/api-[document.version].yaml: Tenant APIs in OpenAPI 3.0 Spec



docs/release-notes.md
# Release Notes

### June 2, 2022

#### Enhancements
- The Email Template for deleted users has been updated.
- Users with an account can now create multiple personal workspaces.
- When a user account is deleted, the user will receive a notification via email the account has been deleted.

#### Fixed
- The content of the Documentation page has been aligned with the page header.

<br>

### May 19, 2022

#### Enhancements
- All page headers have been updated with the same consistent look.
- The ability to set the default example for an endpoint, in the Sandbox on the API Explore page, has been enabled.

#### Fixed
- The direction in which the card carousel is sliding, by the previous and next controls, has been corrected for the Mobile application.

<br>

### May 05, 2022

#### Enhancements
- Banking Hub went full onboard, and lost 'Beta' tag.
- Wiki pages under the [github tenants-doc repository](https://github.com/Fiserv/tenants-doc) has been created to document tenant onboarding.


#### Fixed
- Image rendering on Support pages have been fixed.
- The style for headers and code block has been updated for markdown documentation.

<br>

### April 15, 2022

#### What's new
- The [Release Notes Template](?path=/docs/release-notes-template.md) has been made available for product teams to use.

#### Enhancements
- The [Terms of Use](?path=/docs/terms-of-use/latest.md) has been updated.

#### Fixed
- The mobile layout for cookie banner has been fixed.
- The syntax error in Transaction-Details.md file has been fixed for Commerce Hub [GitHub Issue #128](https://www.github.com/Fiserv/Support/issues/128). 

<br>

### April 1, 2022

#### What's new
- Product & API Versioning has been enabled to view Preview, Active and Previous versions of the product.
- Every product team must add Release Notes as part of the documentation.
- Now Postman collections are downloadable for some products.

#### Enhancements
- When a user is force-logged-out of the system, the user will remain on the same page or be sent to the Home page when the page requires login.
- Elastic Search indexing jobs are now run independently of Mongo DB hashes.
- The documentation tree on the side navigation pane loads near real-time.
- A new job has been created to clean-up the cache on demand for tenants.
- The dark mode switch button has been updated.
- The Runbox on the API explorer page has been updated.
- The Fast Company Award banner has been added to the Home page carousel.
- The Webby Award badge has been added to the footer.
- A feedback clause has been added to Section 3.4 of the [Terms of Use](?path=/docs/terms-of-use/latest.md). 

#### Fixed
- New tab links will now open in a new tab on the API Explorer page.
- A payload generation issue has been corrected [GitHub Issue #115](https://github.com/Fiserv/Support/issues/115).
- The link to Security Questions, in Set Up notifications has been corrected.
- Request Schema properties are now aligned with theh Code Snippet on the API explorer page.
- When the API explorer page or the Documentation page opens from a URL, the left navigation tree will expand with the page link highlighted.

<br>

### March 2022

#### What's new
- Dark mode is available on Developer Studio. Dark mode toggle can be found in the footer

#### Enhancements
- Added pagination for the search results on the Search page
- Added system notifications when data is not available
- Built a new lightweight mock server to support APIs and API versions
- Updated [Terms of Use](?path=/docs/terms-of-use/latest.md) and [Privacy Notice](?path=/docs/privacy-notice/latest.md).
- Updated documentation under Support pages including March Release Notes
- Updated API Explorer page with additional enhancements: download API spec file, code snippet
- Improved APIs & Documentation indexing/crawler to support versioning
- Improved scrollbar consistency across multiple operating systems

#### Fixed
- Fixed pagination for API key table
- Updated error message when creating API key
- Fixed API request/response date format to display correctly in the Runbox on API Explorer page
- Fixed Extended Markdown support to hide HTML comments in the Markdown
- Fixed a bug in navigation for mobile and tablet
- Fixed the visibility for select dropdown in dark mode

<br>

### Feburary 2022

#### Enhancements
- Added additional tags next to a product’s name.
- Updated UI for mobile view of the API Explorer page.

#### Fixed
- Removed added space from one-time password to prevent it from pasting into text field.
- Fixed modals in merchant flow to allow scrolling.

<br>

### January 2022

#### What's new
- Release notes for Developer Studio are now available. Read about the latest feature releases, enhancements, and bug fixes. You can find the release notes on the Support page.  

#### Enhancements
- Expand and view all Merchant IDs and API keys on one page by using the advanced pagination. 

#### Fixed
- Fixed input fields for request schema to display corresponding schema example.
- Fixed dropdowns for request schema to unveil input fields.
- Fixed SwaggerHub example to reflect on Developer Studio.


sample/nodejs/app.js
/**
 * 
 * Copyright (c) 2022 J. Thomas Eck. All rights reserved.
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, this 
 * list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice, 
 * this list of conditions and the following disclaimer in the documentation and/or 
 * other materials provided with the distribution. 
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors may 
 * be used to endorse or promote products derived from this software without specific 
 * prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES 
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AR DISCLAIMED. IN NO EVENT 
 * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 */

const axios = require('axios').default;
const CryptoJS = require('crypto-js')
const Base64 = require('crypto-js/enc-base64');

const CHARGES_BASE_URL = "https://cert.api.fiservapps.com/ch/payments/v1/charges"

// TODO replace these placeholders with the API Key and secret obtained at developer.fiserv.com
const key = "<YOUR API KEY HERE>"
const secret = "<YOUR SECRET HERE>"

// Generate the authorization header value (i.e. transaction signature)
function getSignature(key, secret, data, time, clientRequestId) {
  var rawSignature = key + clientRequestId + time + JSON.stringify(data);

  var computedHash = CryptoJS.algo.HMAC.create(CryptoJS.algo.SHA256, secret.toString());
  computedHash.update(rawSignature);
  computedHash = computedHash.finalize();
  var computedHmac = CryptoJS.enc.Base64.stringify(computedHash)

  return computedHmac
}

// Here's the payload for the Charge request
const chargeData = {
	"amount": {
		"total": 5,
		"currency": "USD"
	},
	"source": {
		"sourceType": "PaymentTrack",
		"encryptionData": {
			"encryptionType": "ON_GUARD",
			"encryptionTarget": "TRACK_2",
			"encryptionBlock": "4614507291879694=078443325742854",
			"keyId": "FFFF109700000E4000340114",
			"deviceType": "INGENICO"
		},
		"pinBlock": {
			"encryptedPin": "0FF7A610CC84CE40",
			"keySerialNumber": "FFFF3D3D3D00232002C9"
		}
	},
	"transactionDetails": {
		"captureFlag": true
	},
	"transactionInteraction": {
		"origin": "POS",
		"posEntryMode": "MAG_STRIPE",
		"posConditionCode": "CARD_PRESENT",
		"terminalTimestamp": "2022-03-10T04:26:56Z",
		"additionalPosInformation": {
			"dataEntrySource": "MOBILE_TERMINAL",
			"posFeatures": {
				"pinAuthenticationCapability": "CAN_ACCEPT_PIN",
				"terminalEntryCapability": "MAG_STRIPE_MANUAL_CHIP"
			}
		}
	},
	"merchantDetails": {
		"merchantId": "100009000000035",
		"terminalId": "10000002"
	},
	"additionalDataCommon": {
		"directedRouting": {
			"processors": [{
				"code": "NASHVILLE",
				"platform": "NORTH",
				"priority": "PRIMARY"
			}]
		}
	}
}

// Here's the payload for the Cancel request
const cancelData = {
  "amount": {
    "total": 5,
    "currency": "USD"
  },
  "merchantDetails": {
		"merchantId": "100009000000035",
		"terminalId": "10000002"
  }
}

function doCancel(transactionId) {

	console.log("Performing Cancel transaction")

	var time = new Date().getTime()
	var clientRequestId = (Math.random() * 10000000) + 1
	var signature = getSignature(key, secret, cancelData, time, clientRequestId)

	var headers = {
		'Accept' : 'application/json',
		'Content-Type': 'application/json',
		'Accept-language' : 'en',
		'Auth-Token-Type' : 'HMAC',
		'Timestamp' : time,
		'Api-Key' : key,
		'Client-Request-Id' : clientRequestId,
		'Authorization': signature,
	}

	var cancel_url = CHARGES_BASE_URL + "/" + transactionId + "/cancel"
	console.log("Cancel URL is", cancel_url)
	axios.post(cancel_url, cancelData, {
			headers:headers
		})
		.then((response) => {
				//console.log(response)
				transactionId = response.data.gatewayResponse.transactionProcessingDetails.transactionId
				console.log("Cancel was successful. Transaction Id is", transactionId)
		})
		.catch(err => {
				console.log("Error performing cancel", err.response);
		})
}

function doCharge() {

	console.log("Performing Charge transaction")

	var time = new Date().getTime()
	var clientRequestId = (Math.random() * 10000000) + 1
	var signature = getSignature(key, secret, chargeData, time, clientRequestId)
	var transactionId = ""

	var headers = {
	  'Accept' : 'application/json',
	  'Content-Type': 'application/json',
	  'Accept-language' : 'en',
	  'Auth-Token-Type' : 'HMAC',
	  'Timestamp' : time,
	  'Api-Key' : key,
	  'Client-Request-Id' : clientRequestId,
	  'Authorization': signature,
	}

	// Perform the authorization
	axios.post(CHARGES_BASE_URL, chargeData, {
	  	headers:headers
	  })
	  .then((response) => {
	      //console.log(response)
				transactionId = response.data.gatewayResponse.transactionProcessingDetails.transactionId
				console.log("Charge was successful.  Transaction Id is", transactionId)

				// Cancel that auth
				doCancel(transactionId)
		  })
		  .catch(err => {
		      console.log("Error performing charge transaction", err.response);
		  })
}

// Vefify that API Key and secret have been provided
if( key === "<YOUR API KEY HERE>" || secret === "<YOUR SECRET HERE>") {
  console.log("Please replace the API Key and secret with the values obtained at developer.fiserv.com")
  process.exit(1);
}

// Perform a charge and then an immediate cancel
doCharge()


sample/golang/app.go
/**
 * 
 * Copyright (c) 2022 J. Thomas Eck. All rights reserved.
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, this 
 * list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice, 
 * this list of conditions and the following disclaimer in the documentation and/or 
 * other materials provided with the distribution. 
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors may 
 * be used to endorse or promote products derived from this software without specific 
 * prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES 
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AR DISCLAIMED. IN NO EVENT 
 * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 */

package main

import (
	//"flag"
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	b64 "encoding/base64"
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/tidwall/gjson"
)

const CHARGES_BASE_URL = "https://cert.api.fiservapps.com/ch/payments/v1/charges"

// TODO -- replace placeholders with the API Key and secret obtained at developer.fiserv.com
const key = "<YOUR API KEY HERE>"
const secret = "<YOUR SECRET HERE>"

// Here are the bodies of the requests that we'll be sending
const gChargeData = `{
  "amount": {
    "total": 500,
    "currency": "USD"
  },
  "source": {
    "sourceType": "PaymentTrack",
    "encryptionData": {
      "encryptionType": "ON_GUARD",
      "encryptionTarget": "TRACK_2",
      "encryptionBlock": "4614507291879694=078443325742854",
      "keyId": "FFFF109700000E4000340114",
      "deviceType": "INGENICO"
    },
    "pinBlock": {
      "encryptedPin": "0FF7A610CC84CE40",
      "keySerialNumber": "FFFF3D3D3D00232002C9"
    }
  },
  "transactionDetails": {
    "captureFlag": true
  },
  "transactionInteraction": {
    "origin": "POS",
    "posEntryMode": "MAG_STRIPE",
    "posConditionCode": "CARD_PRESENT",
    "terminalTimestamp": "2022-03-10T04:26:56Z",
    "additionalPosInformation": {
      "dataEntrySource": "MOBILE_TERMINAL",
      "posFeatures": {
        "pinAuthenticationCapability": "CAN_ACCEPT_PIN",
        "terminalEntryCapability": "MAG_STRIPE_MANUAL_CHIP"
      }
    }
  },
  "merchantDetails": {
    "merchantId": "100009000000035",
    "terminalId": "10000002"
  },
  "additionalDataCommon": {
    "directedRouting": {
      "processors": [
        {
          "code": "NASHVILLE",
          "platform": "NORTH",
          "priority": "PRIMARY"
        }
      ]
    }
  }
}`

const gCancelData = `{
  "amount": {
    "total": 500,
    "currency": "USD"
  },
  "merchantDetails": {
    "merchantId": "100009000000035",
    "terminalId": "10000002"
  }
}`

// Will be set in the charge request, and read in the cancel request
var gTransactionId = ""

//	getSignature is used to generate the HMAC signature for the request.
func getSignature(key string, secret string, data string, timeStamp int64, clientRequestId int) string {

	rawSignature := key + fmt.Sprint(clientRequestId) + fmt.Sprint(timeStamp) + data

	// Create a new HMAC by defining the hash type and the key (as byte array)
	h := hmac.New(sha256.New, []byte(secret))

	// Write Data to it
	h.Write([]byte(rawSignature))

	signature := h.Sum(nil)
	hexSignature := hex.EncodeToString(signature) //	this step was missing
	sha := b64.StdEncoding.EncodeToString([]byte(hexSignature))

	return sha
}

//	cleanse removes leading and trailing \" from a string
func cleanse(input string) string {
	return strings.Trim(input, "\"")
}

//	sendChargeRequest is used to send a Charge request to Commerce Hub. If successful,
//	the transaction ID of the Charge is stored in gTransactionId.
func sendChargeRequest() {

	// Create an http client
	client := http.DefaultClient

	//	Compute header values
	timeStamp := time.Now().UnixNano() / int64(time.Millisecond)

	// Generate a nice random number (seeded w/current time) for idempotency check
	randSource := rand.NewSource(timeStamp)
	clientRequestId := rand.New(randSource).Intn(10000000) + 1

	signature := getSignature(key, secret, gChargeData, timeStamp, clientRequestId)

	// Setup http request
	req, err := http.NewRequest("POST", CHARGES_BASE_URL, bytes.NewBuffer([]byte(gChargeData)))
	log.Println("Charge URL is", CHARGES_BASE_URL)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Accept-language", "en")
	req.Header.Add("Auth-Token-Type", "HMAC")
	req.Header.Add("Timestamp", strconv.Itoa(int(timeStamp)))
	req.Header.Add("Api-Key", key)
	req.Header.Add("Client-Request-Id", strconv.Itoa(int(clientRequestId)))
	req.Header.Add("Authorization", signature)

	// Make http call
	log.Println("Sending Charge request")

	resp, err := client.Do(req)
	if err != nil || (resp.StatusCode != 201 && resp.StatusCode != 200) {
		// handle error
		log.Println("Error invoking charges endpoint:", resp.StatusCode)
	} else {
		defer resp.Body.Close()
		body, _ := ioutil.ReadAll(resp.Body)
		//log.Print(string(body))

		jsonResp := gjson.Parse(string(body))
		gTransactionId = cleanse(jsonResp.Get("gatewayResponse.transactionProcessingDetails.transactionId").Raw)
		log.Println("Charge request successful, transactionId is", gTransactionId)
	}
}

//	sendCancelRequest is used to send a Cancel request to Commerce Hub. It uses the transaction ID
//	created by sendChargeRequest to identify the cancelled transaction.
func sendCancelRequest() {

	// Create an http client
	client := http.DefaultClient

	timeStamp := time.Now().UnixNano() / int64(time.Millisecond)

	// Generate a nice random number (seeded w/current time) for idempotency check
	randSource := rand.NewSource(timeStamp)
	clientRequestId := rand.New(randSource).Intn(10000000) + 1

	signature := getSignature(key, secret, gCancelData, timeStamp, clientRequestId)

	// Setup http request
	cancelUrl := CHARGES_BASE_URL + "/" + gTransactionId + "/cancel"
	log.Println("Cancel URL is", cancelUrl)
	req, err := http.NewRequest("POST", cancelUrl, bytes.NewBuffer([]byte(gCancelData)))
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Accept-language", "en")
	req.Header.Add("Auth-Token-Type", "HMAC")
	req.Header.Add("Timestamp", strconv.Itoa(int(timeStamp)))
	req.Header.Add("Api-Key", key)
	req.Header.Add("Client-Request-Id", strconv.Itoa(int(clientRequestId)))
	req.Header.Add("Authorization", signature)

	// Make http call
	log.Println("Sending Cancel request")

	resp, err := client.Do(req)
	if err != nil || (resp.StatusCode != 201 && resp.StatusCode != 200) {
		// handle error
		log.Println("Error invoking cancel endpoint:", resp.StatusCode)
	} else {
		defer resp.Body.Close()
		body, _ := ioutil.ReadAll(resp.Body)
		//log.Print(string(body))

		jsonResp := gjson.Parse(string(body))
		// Set the global transactionId so we can cancel it next
		gTransactionId = cleanse(jsonResp.Get("gatewayResponse.transactionProcessingDetails.transactionId").Raw)
		log.Println("Cancel request successful, transactionId is", gTransactionId)
	}
}

func main() {

	// Verify that API Key and Secret have been supplied
	if key == "<YOUR API KEY HERE>" || secret == "<YOUR SECRET HERE>" {
		log.Println("Please replace the placeholders with the API Key and Secret obtained at developer.fiserv.com")
		return
	}

	sendChargeRequest()
	if gTransactionId != "" {
		sendCancelRequest()
	}
}
